src/test/java/com/google/sps/servlets/UserDataServletTest.java

```diff
@@ -0,0 +1,172 @@
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+import ch.qos.logback.classic.LoggerContext;
+
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.sps.data.Email;
+import com.google.sps.api.GmailConfiguration;
+import com.google.sps.MemoryAppender;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.slf4j.LoggerFactory;
+
+@RunWith(JUnit4.class)
+public final class UserDataServletTest {
+
+  private static final String LOGGER_NAME = "com.google.sps.servlets";
+  private static final String SUB_SUCCESS_MSG = 
+    "User was subscribed to email notifications and added to Datastore.";
+  private static final String UNSUB_SUCCESS_MSG = 
+    "User was unsubscribed and removed from Datastore.";
+  private static final String UNSUB_FAIL_MSG = 
+    "User was unable to unsubscribe but was never subscribed.";
+
+  private static final String NAME = "test";
+  private static final String EMAIL_A = "testa@google.com";
+  private static final String EMAIL_B = "testb@google.com";
+  private static final String COLLEGE_A = "000001";
+  private static final String COLLEGE_B = "000002";
+  private static final String UNSUB = "unsubscribe";
+  private static final String SUB = "subscribe";
+
+  @Mock private static HttpServletRequest mRequest;
+  @Mock private static HttpServletResponse mResponse;
+  @Mock private static GmailConfiguration mGmailConfiguration;
+
+  private static UserDataServlet userDataServlet;
+  private static MemoryAppender memoryAppender;
+  private static LocalServiceTestHelper helper =
+    new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Before
+  public void setUp() {
+    
+    helper.setUp();
+
+    Logger logger = (Logger) LoggerFactory.getLogger(LOGGER_NAME);
+    memoryAppender = new MemoryAppender();
+    memoryAppender.setContext((LoggerContext) LoggerFactory.getILoggerFactory());
+    logger.setLevel(Level.DEBUG);
+    logger.addAppender(memoryAppender);
+    memoryAppender.start();
+    userDataServlet = new UserDataServlet();
+
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @After
+  public void tearDown() {
```

Comment:
Good job for making sure to teardown tests

Reply:
Thank you!